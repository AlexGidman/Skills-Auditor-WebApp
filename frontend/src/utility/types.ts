export const SYSTEM_ROLES = ["StaffUser", "Manager", "Admin"];
export const STAFF_USER = SYSTEM_ROLES[0];
export const MANAGER_SR = SYSTEM_ROLES[1];
export const ADMIN = SYSTEM_ROLES[2];

export const JOB_ROLES = ["Mid-Level Developer", "Senior Developer", "Manager"];
export const MIDLEVEL_DEVELOPER = JOB_ROLES[0];
export const SENIOR_DEVELOPER = JOB_ROLES[1];
export const MANAGER_JR = JOB_ROLES[2];

export const TOKEN_COOKIE = "token";

// Week in seconds
export const TOKEN_LIFE = 60 * 60 * 24 * 7;

export const SKILL_LEVELS = ["None", "Basic", "Intermediate", "High", "Expert"];
export const NONE = SKILL_LEVELS[0];
export const BASIC = SKILL_LEVELS[1];
export const INTERMEDIATE = SKILL_LEVELS[2];
export const HIGH = SKILL_LEVELS[3];
export const EXPERT = SKILL_LEVELS[4];

export type User = {
    firstName: string;
    lastName: string;
    email: string;
    password: string;
};
