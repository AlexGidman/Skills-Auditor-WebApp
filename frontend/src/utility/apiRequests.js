import axios from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders, id) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders, name) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (authHeaders, id, name) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders, id) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (authHeaders, name, category_id) => {
    const data = {
        name,
        category_id,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (authHeaders, id, name, categoryId) => {
    const data = {
        id,
        name,
        category_id: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders, id) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders,
    id,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        id,
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders, id) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (authHeaders, userId, reportId) => {
    const data = {
        user_id: userId,
        report_id: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders,
    user_id,
    skill_id,
    skill_level,
    notes,
    expiry_date,
) => {
    const data = {
        user_id,
        skill_id,
        skill_level,
        notes,
        expiry_date,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (authHeaders, id, skill_level, notes) => {
    const data = {
        id,
        skill_level,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email, password) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
