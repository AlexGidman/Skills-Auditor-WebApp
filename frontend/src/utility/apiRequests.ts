import axios, { AxiosRequestHeaders } from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders: AxiosRequestHeaders, name: string) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (
    authHeaders: AxiosRequestHeaders,
    name: string,
    categoryId: string,
) => {
    const data = {
        name,
        categoryId,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
    categoryId: string,
) => {
    const data = {
        id,
        name,
        categoryId: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        id,
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders: AxiosRequestHeaders,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    reportId: string,
) => {
    const data = {
        userId: userId,
        reportId: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    skillId: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
    expiryDate: Date,
) => {
    const data = {
        userId,
        skillId,
        skillLevel,
        notes,
        expiryDate,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
) => {
    const data = {
        id,
        skillLevel,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email: string, password: string) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
