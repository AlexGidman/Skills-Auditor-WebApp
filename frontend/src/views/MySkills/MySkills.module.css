.header {
    margin-top: 10%;
}

.button {
    margin-top: var(--s2);
    /* flex: 44%; */
    /* min-width: 14rem; */
}

.buttonWrapper {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    flex-wrap: wrap;
}

.form {
    margin: 0 20%;
}

.welcomeMessage {
    font-size: xx-large;
}

.listview {
    margin-bottom: var(--s2);
    margin-left: auto;
    margin-right: auto;
}

.table {
    width: 100%;
    margin-left: auto;
    margin-right: auto;
}

.rowHeightColSpan {
    height: 6.25rem;
}

@media (max-width: 900px) {
    .form {
        margin: 0;
    }
    .button {
        flex: 100%;
        margin-left: unset;
        margin-right: unset;
    }
}
