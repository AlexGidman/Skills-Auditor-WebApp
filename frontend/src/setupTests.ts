// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
