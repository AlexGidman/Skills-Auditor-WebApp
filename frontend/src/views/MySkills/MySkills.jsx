import React, { useState, useEffect } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { LinkButton } from "../../components/LinkButton/LinkButton";
import { Button } from "../../components/Button/Button";
import {
    getSelectOptionsFromArray,
    useAPI,
    formatDate,
    formatSkillLevel,
} from "../../utility/helper";
import { getAllStaffSkills, deleteStaffSkill } from "../../utility/apiRequests";
import styles from "./MySkills.module.css";
import { ListView } from "../../components/ListView/ListView";

export const MySkills = () => {
    const [currentUser] = useOutletContext();
    const { data, loading, error } = useAPI(getAllStaffSkills, [currentUser.id]);
    if (loading) return <LoadingPlaceholder />;

    const sortedSkills = data && data.sort((a, b) => a.skillID.name.localeCompare(b.skillID.name));

    return (
        <>
            <header>
                <h1>My Skills</h1>
            </header>
            <section>
                <Form data={sortedSkills} dataError={error} />
            </section>
        </>
    );
};

const Form = ({ data, dataError }) => {
    const navigate = useNavigate();
    const [, setShowToast] = useOutletContext();
    const [selectedSkill, setSelectedSkill] = useState(data ? data[0].id : null);
    const [skillLevel, setSkillLevel] = useState(
        data ? formatSkillLevel(data[0].skill_level) : null,
    );
    const [expiryDate, setExpiryDate] = useState(data ? formatDate(data[0].expiry_date) : null);
    const [skillNotes, setSkillNotes] = useState(data ? data[0].notes : "");

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        error,
        loading,
        callback: deleteSkill,
    } = useAPI(deleteStaffSkill, [selectedSkill], true);

    useEffect(() => {
        if (navigateBack && !loading) {
            error
                ? setShowToast({ error: "Failed to delete user skill" })
                : setShowToast({ success: "Successfully deleted user skill" });
            navigate(0);
        }
    }, [error, loading, navigate, navigateBack, setShowToast]);

    useEffect(() => {
        setFormChanged(selectedSkill !== undefined);
    }, [selectedSkill]);

    return (
        <form className={styles.form}>
            {data && (
                <>
                    <ListView
                        className={styles.listview}
                        options={getSelectOptionsFromArray(
                            data.map((userSkill) => {
                                return userSkill.id;
                            }),
                            data.map((userSkill) => {
                                return `${userSkill.skillID.name} - ${userSkill.skillID.category.name}`;
                            }),
                        )}
                        value={selectedSkill}
                        labelText="Select a user skill"
                        onChange={(e) => {
                            let resultArray = data.filter((x) => x.id == e.target.value);
                            setSelectedSkill(e.target.value);
                            setSkillLevel(formatSkillLevel(resultArray[0].skill_level));
                            setExpiryDate(formatDate(resultArray[0].expiry_date));
                            setSkillNotes(resultArray[0].notes);
                        }}
                    />

                    <table className={styles.table}>
                        <tbody>
                            <tr>
                                <th>Skill Level:</th>
                                <td>{skillLevel}</td>
                            </tr>
                            <tr>
                                <th>Expiry Date:</th>
                                <td>{expiryDate}</td>
                            </tr>
                        </tbody>
                    </table>

                    <table className={styles.table}>
                        <tbody>
                            <tr className={styles.rowHeightColSpan}>
                                <td>{skillNotes}</td>
                            </tr>
                        </tbody>
                    </table>
                </>
            )}
            {dataError && <ErrorMessage error={dataError} />}

            <div className={styles.buttonWrapper}>
                <LinkButton
                    className={styles.button}
                    disabled={!selectedSkill}
                    name="Edit"
                    path={`/edit/staffskill/${selectedSkill}`}
                />
                <Button
                    className={styles.button}
                    loading={loading}
                    disabled={!formChanged || !selectedSkill}
                    onClick={(e) => {
                        e.preventDefault();
                        deleteSkill();
                        setNavigateBack(true);
                    }}
                >
                    Delete
                </Button>
                <LinkButton className={styles.button} name="Add Skill" path={`/addstaffskill`} />
            </div>
        </form>
    );
};
