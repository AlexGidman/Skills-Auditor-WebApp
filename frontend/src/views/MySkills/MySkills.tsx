import React, { useState, useEffect } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { LinkButton } from "../../components/LinkButton/LinkButton";
import { Button } from "../../components/Button/Button";
import {
    getSelectOptionsFromArray,
    useAPI,
    formatDate,
    formatSkillLevel,
} from "../../utility/helper";
import { getAllStaffSkills, deleteStaffSkill } from "../../utility/apiRequests";
import styles from "./MySkills.module.css";
import { ListView } from "../../components/ListView/ListView";
import { StaffSkill } from "../../utility/types";

export const MySkills = () => {
    // @ts-ignore TODO fix type for AppOutletContext here
    const [currentUser] = useOutletContext();
    const { data, loading, error } = useAPI<StaffSkill[]>(getAllStaffSkills, [currentUser.id]);
    if (loading) return <LoadingPlaceholder />;

    const sortedSkills = data
        ? data.sort((a, b) => a.skill.name.localeCompare(b.skill.name))
        : null;
    return (
        <>
            <header>
                <h1>My Skills</h1>
            </header>
            <section>
                <Form data={sortedSkills} dataError={error} />
            </section>
        </>
    );
};

interface FormProps {
    data: StaffSkill[] | null;
    dataError: Error | null;
}

const Form = ({ data, dataError }: FormProps) => {
    const navigate = useNavigate();
    // @ts-ignore TODO fix type for AppOutletContext here
    const [, setShowToast] = useOutletContext();
    const [selectedSkill, setSelectedSkill] = useState(data ? data[0].id : "");
    const [skillLevel, setSkillLevel] = useState(
        data ? formatSkillLevel(data[0].skillLevel) : null,
    );
    const [expiryDate, setExpiryDate] = useState(data ? formatDate(data[0].expiryDate) : null);
    const [skillNotes, setSkillNotes] = useState(data ? data[0].notes : "");

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        error,
        loading,
        callback: deleteSkill,
    } = useAPI(deleteStaffSkill, [selectedSkill], true);

    useEffect(() => {
        if (navigateBack && !loading) {
            error
                ? setShowToast({ error: "Failed to delete user skill" })
                : setShowToast({ success: "Successfully deleted user skill" });
            navigate(0);
        }
    }, [error, loading, navigate, navigateBack, setShowToast]);

    useEffect(() => {
        setFormChanged(selectedSkill !== undefined);
    }, [selectedSkill]);

    return (
        <form className={styles.form}>
            {data && (
                <>
                    <ListView
                        className={styles.listview}
                        options={getSelectOptionsFromArray(
                            data.map((userSkill) => {
                                return userSkill.id;
                            }),
                            data.map((userSkill) => {
                                return `${userSkill.skill.name} - ${userSkill.skill.category.name}`;
                            }),
                        )}
                        value={selectedSkill}
                        labelText="Select a user skill"
                        onChange={(e) => {
                            let resultArray = data.filter((x) => x.id == e.target.value);
                            setSelectedSkill(e.target.value);
                            setSkillLevel(formatSkillLevel(resultArray[0].skillLevel));
                            setExpiryDate(formatDate(resultArray[0].expiryDate));
                            setSkillNotes(resultArray[0].notes);
                        }}
                    />

                    <table className={styles.table}>
                        <tbody>
                            <tr>
                                <th>Skill Level:</th>
                                <td>{skillLevel}</td>
                            </tr>
                            <tr>
                                <th>Expiry Date:</th>
                                <td>{expiryDate}</td>
                            </tr>
                        </tbody>
                    </table>

                    <table className={styles.table}>
                        <tbody>
                            <tr className={styles.rowHeightColSpan}>
                                <td>{skillNotes}</td>
                            </tr>
                        </tbody>
                    </table>
                </>
            )}
            {dataError && <ErrorMessage error={dataError} />}

            <div className={styles.buttonWrapper}>
                <LinkButton
                    className={styles.button}
                    disabled={!selectedSkill}
                    name="Edit"
                    path={`/edit/staffskill/${selectedSkill}`}
                />
                <Button
                    className={styles.button}
                    loading={loading}
                    disabled={!formChanged || !selectedSkill}
                    onClick={(e) => {
                        e.preventDefault();
                        deleteSkill();
                        setNavigateBack(true);
                    }}
                >
                    Delete
                </Button>
                <LinkButton className={styles.button} name="Add Skill" path={`/addstaffskill`} />
            </div>
        </form>
    );
};
