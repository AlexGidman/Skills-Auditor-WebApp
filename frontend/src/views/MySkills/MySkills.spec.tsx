import { render, screen } from "@testing-library/react";
import { MySkills } from "./MySkills";
import { mockAdminUser, mockApiRequests, mockApiResponse, mockError } from "../../setupTests";
import * as ReactDom from "react-router-dom";
import React from "react";
import { Category, Skill, StaffSkill } from "../../utility/types";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
    };
});

const mockCategory: Category = { id: "1", name: "testname1" };

const mockSkill1: Skill = {
    id: "1",
    name: "Skill1",
    category: mockCategory,
    categoryId: mockCategory.id,
};

const mockStaffSkill1: StaffSkill = {
    id: "100",
    skillId: mockSkill1.id,
    skill: mockSkill1,
    skillLevel: "1",
    userId: mockAdminUser.id,
    notes: "test notes",
    expiryDate: "0000-00-00",
};

describe("MySkills", () => {
    it("renders correctly when data successful", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getAllStaffSkills.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: [mockStaffSkill1] }),
        );

        mockApiRequests.deleteStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: "Success" }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <MySkills />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "My Skills" })).toBeInTheDocument();

        expect(screen.getByRole("listbox", { name: "Select a user skill" })).toBeInTheDocument();
        expect(
            screen.getByRole("option", {
                name: `${mockStaffSkill1.skill.name} - ${mockStaffSkill1.skill.category.name}`,
            }),
        ).toBeInTheDocument();

        expect(screen.getAllByRole("table").length).toBe(2);
        expect(screen.getByRole("columnheader", { name: "Skill Level:" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: "None" })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Expiry Date:" })).toBeInTheDocument();

        expect(screen.getByRole("cell", { name: mockStaffSkill1.notes })).toBeInTheDocument();

        expect(screen.getByRole("link", { name: "Edit" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Delete" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Delete" })).toBeDisabled();
        expect(screen.getByRole("link", { name: "Add Skill" })).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("renders correctly when data retrieval fails", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });
        mockApiRequests.getAllStaffSkills.mockImplementation(() => Promise.reject(mockError));
        mockApiRequests.deleteStaffSkill.mockImplementation(() => Promise.reject(mockError));
        const { container } = render(
            <ReactDom.MemoryRouter>
                <MySkills />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "My Skills" })).toBeInTheDocument();

        expect(
            screen.queryByRole("listbox", { name: "Select a user skill" }),
        ).not.toBeInTheDocument();

        expect(screen.queryByRole("table")).not.toBeInTheDocument();

        expect(screen.getByRole("link", { name: "Edit" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Delete" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Delete" })).toBeDisabled();
        expect(screen.getByRole("link", { name: "Add Skill" })).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
