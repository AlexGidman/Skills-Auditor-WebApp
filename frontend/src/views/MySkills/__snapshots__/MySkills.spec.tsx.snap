// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MySkills renders correctly when data retrieval fails 1`] = `
<div>
  <header>
    <h1>
      My Skills
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <div
        aria-label="Oops"
        class="component"
        role="alert"
      >
        <h1>
          Something went wrong...
        </h1>
        <p>
          Oops
        </p>
      </div>
      <div
        class="buttonWrapper"
      >
        <a
          class="component button disabled"
          href="/edit/staffskill/"
        >
          Edit
        </a>
        <button
          class="component button"
          disabled=""
        >
          Delete
        </button>
        <a
          class="component button"
          href="/addstaffskill"
        >
          Add Skill
        </a>
      </div>
    </form>
  </section>
</div>
`;

exports[`MySkills renders correctly when data successful 1`] = `
<div>
  <header>
    <h1>
      My Skills
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <div
        class="component component listview"
      >
        <label
          class="label"
          for="Select-a user skill-undefined"
        >
          Select a user skill
          :
        </label>
        <select
          aria-label="Select a user skill"
          class="select"
          name="Select a user skill"
          size="5"
        >
          <option
            value="100"
          >
            Skill1 - testname1
          </option>
        </select>
      </div>
      <table
        class="table"
      >
        <tbody>
          <tr>
            <th>
              Skill Level:
            </th>
            <td>
              None
            </td>
          </tr>
          <tr>
            <th>
              Expiry Date:
            </th>
            <td>
              N/A
            </td>
          </tr>
        </tbody>
      </table>
      <table
        class="table"
      >
        <tbody>
          <tr
            class="rowHeightColSpan"
          >
            <td>
              test notes
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="buttonWrapper"
      >
        <a
          class="component button"
          href="/edit/staffskill/100"
        >
          Edit
        </a>
        <button
          class="component button"
          disabled=""
        >
          Delete
        </button>
        <a
          class="component button"
          href="/addstaffskill"
        >
          Add Skill
        </a>
      </div>
    </form>
  </section>
</div>
`;
